docs/core/v3.0.rst

```diff
@@ -0,0 +1,1762 @@
+========================
+
+This section introduces and defines some key terms and explains the
+conceptual model underpinning the Zarr format.
+
+.. _hierarchy:
+
+*Hierarchy*
+
+    A Zarr hierarchy is a tree structure, where each node in the tree
+    is either a group_ or an array_. Group nodes may have children but
+    array nodes may not. All nodes in a hierarchy have a name_ and a
+    path_.
+
+.. _group:
+.. _groups:
+
+*Group*
+
+    A group is a node in a hierarchy_ that may have child nodes.
+
+.. _array:
+.. _arrays:
+
+*Array*
+
+    An array is a node in a hierarchy_. An array is a data structure
+    with zero or more dimensions_ whose lengths define the shape_ of
+    the array. An array contains zero or more data elements_. All
+    elements_ in an array conform to the same `data type`_. An array
+    may not have child nodes.
+
+.. _name:
+.. _names:
+
+*Name*
+
+    Each node in a hierarchy_ has a name, which is a string of
+    characters with some additional constraints defined in the section
+    on `node names`_ below. Two sibling nodes cannot have the same
+    name. The root node does not have a name.
+
+.. _path:
+.. _paths:
+
+*Path*
+
+    Each node in a hierarchy_ has a path which uniquely identifies
+    that node and defines its location within the hierarchy_. The path
+    is formed by joining together the "/" character, followed by the
+    name_ of each ancestor node separated by the "/" character,
+    followed by the name_ of the node itself. For example, the path
+    "/foo/bar" identifies a node named "bar", whose parent is named
+    "foo", whose parent is the root of the hierarchy. The path "/"
+    identifies the root node.
+
+    A path always starts with ``/``.
+
+    QUESTION: do we want to codify that group ``path`` end, can end, or must end
+    in ``/`` ?
+
+.. _dimension:
+.. _dimensions:
+
+*Dimension*
+
+    An array_ has a fixed number of zero or more dimensions. Each
+    dimension has an integer length. This specification only considers
+    the case where the lengths of all dimensions are finite. However,
+    `extensions`_ may be defined which allow a dimension to have
+    an infinite or variable length.
+
+.. _shape:
+
+*Shape*
+
+    The shape of an array_ is the tuple of dimension_ lengths. For
+    example, if an array_ has 2 dimensions_, where the length of the
+    first dimension_ is 100 and the length of the second dimension_ is
+    20, then the shape of the array_ is (100, 20). A shape can be the empty
+    tuple in the case of zero-dimension arrays (scalar)
+
+.. _element:
+.. _elements:
+
+*Element*
+
+    An array_ contains zero or more elements. Each element can be
+    identified by a tuple of integer coordinates, one for each
+    dimension_ of the array_. If all dimensions_ of an array_ have
+    finite length, then the number of elements in the array_ is given
+    by the product of the dimension_ lengths. An array_ may not have
+    been fully initialized.
```

Comment:
This last sentence feels slightly out of place -- is it possible that the concept of array initialization would be more appropriately introduced in the section about data types?

Reply:
Agreed, any comments about initialization should be specified when discussing the fill value.

Reply:
Since this is the `Concepts and terminology` section this seems too detailed to me too, I removed it in 0a60d82 of PR #171.